File: python system/inventory_system/models/audit_log.py
from enum import Enum
from datetime import datetime
from typing import Optional, Dict, Any


class AuditAction(Enum):
    CREATED = "created"
    UPDATED = "updated"
    DELETED = "deleted"
    STOCK_IN = "stockIn"
    STOCK_OUT = "stockOut"
    ADJUSTMENT = "adjustment"


class AuditLog:
    def __init__(
        self,
        id: str,
        timestamp: datetime,
        action: AuditAction,
        entity_type: str,
        entity_id: str,
        user_id: Optional[str] = None,
        before_snapshot: Optional[str] = None,
        after_snapshot: Optional[str] = None,
        notes: Optional[str] = None,
    ):
        self.id = id
        self.timestamp = timestamp
        self.action = action
        self.entity_type = entity_type
        self.entity_id = entity_id
        self.user_id = user_id
        self.before_snapshot = before_snapshot
        self.after_snapshot = after_snapshot
        self.notes = notes

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "timestamp": self.timestamp.isoformat(),
            "action": self.action.value,
            "entity_type": self.entity_type,
            "entity_id": self.entity_id,
            "user_id": self.user_id,
            "before_snapshot": self.before_snapshot,
            "after_snapshot": self.after_snapshot,
            "notes": self.notes,
        }

    def __str__(self) -> str:
        return f"AuditLog(id={self.id}, action={self.action.value}, entity={self.entity_type}:{self.entity_id})"

File: dart/transaction_system/dart_application_1/bin/audit_log.dart
// lib/models/audit_log.dart

enum AuditAction {
  created,
  updated,
  deleted,
  stockIn,
  stockOut,
  adjustment,
}

class AuditLog {
  final String id;
  final DateTime timestamp;
  final AuditAction action;
  final String entityType;   // e.g. 'InventoryItem', 'Transaction'
  final String entityId;
  final String? userId;
  final String? beforeSnapshot; // JSON string of entity before change
  final String? afterSnapshot;  // JSON string of entity after change
  final String? notes;

  const AuditLog({
    required this.id,
    required this.timestamp,
    required this.action,
    required this.entityType,
    required this.entityId,
    this.userId,
    this.beforeSnapshot,
    this.afterSnapshot,
    this.notes,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'timestamp': timestamp.toIso8601String(),
        'action': action.name,
        'entity_type': entityType,
        'entity_id': entityId,
        'user_id': userId,
        'before_snapshot': beforeSnapshot,
        'after_snapshot': afterSnapshot,
        'notes': notes,
      };

  factory AuditLog.fromJson(Map<String, dynamic> json) => AuditLog(
        id: json['id'] as String,
        timestamp: DateTime.parse(json['timestamp'] as String),
        action: AuditAction.values.byName(json['action'] as String),
        entityType: json['entity_type'] as String,
        entityId: json['entity_id'] as String,
        userId: json['user_id'] as String?,
        beforeSnapshot: json['before_snapshot'] as String?,
        afterSnapshot: json['after_snapshot'] as String?,
        notes: json['notes'] as String?,
      );

  @override
  String toString() =>
      'AuditLog(id: $id, action: ${action.name}, entity: $entityType[$entityId])';
}
